File: Linked list/Linked List Cycle II/Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode * tail = new ListNode(-1);
        ListNode * p = head;
        unordered_set<ListNode*> sets;
        while(true){
            if(sets.size() == 0)
                sets.insert(p);
            else if(sets.find(p) != sets.end())
                return p;
            else
                sets.insert(p);
            if(p == NULL) return p;
            p = p->next;
        }
    }
};

File: Linked list/Linked List Cycle II/Linked List Cycle II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode p = head;
        Set<ListNode> sets = new HashSet<ListNode> ();
        while(true){
            if(sets.size() == 0)
                sets.add(p);
            else if(sets.contains(p) == true)
                return p;
            else
                sets.add(p);
            if(p == null) return p;
            p = p.next;
        }
    }
}
